File: frontend/types/asistencia.ts
export type TipoActividad =
    | "analisis_partido"
    | "preparacion_fisica"
    | "reunion_ordinaria"
    | "reunion_extraordinaria"

export type EstadoAsistencia =
    | "presente"
    | "ausente"
    | "justificado"
    | "licencia"
    | "tardanza"
    | "justificacion"

// Tipo de día para asistencia
export type TipoDia = "OBLIGATORIO" | "OPCIONAL" | "DESCANSO"

export interface Arbitro {
    id: string
    codigoCODAR: string
    nombres: string
    apellidoPaterno: string
    apellidoMaterno: string
    categoria: string
    telefono: string
    email?: string
}

export interface AsistenciaArbitro {
    arbitroId: string
    estado: EstadoAsistencia
    horaRegistro: string
    observaciones: string
}

export interface RegistroAsistencia {
    id: string
    fecha: string
    horaInicio: string
    horaFin: string
    tipoActividad: TipoActividad
    descripcion: string
    ubicacion: string
    responsable: string
    arbitros: AsistenciaArbitro[]
    createdAt: string
}

// ========== NUEVOS TIPOS PARA MEJORA DE ASISTENCIA ==========

// Información del día actual
export interface DiaInfo {
    fecha: string
    diaSemana: number
    nombreDia: string
    esObligatorio: boolean
    tipoDia: TipoDia
}

// Estadísticas por día de la semana
export interface EstadisticaDia {
    dia: string
    numeroDia: number
    esObligatorio: boolean
    total: number
    presentes: number
    ausentes: number
    tardanzas: number
    justificaciones: number
    porcentajeAsistencia: number
}

export interface EstadisticasPeriodo {
    periodo: {
        inicio: string
        fin: string
    }
    porDia: Record<string, EstadisticaDia>
    resumen: {
        totalRegistros: number
        presentes: number
        porcentajeGeneral: number
    }
}

export interface EstadisticasDiasObligatorios {
    periodo: {
        inicio: string
        fin: string
    }
    diasObligatorios: {
        total: number
        presentes: number
        ausentes: number
        tardanzas: number
        justificaciones: number
        porcentajeAsistencia: number
    }
}

// Request para registro con retraso
export interface RegistroConRetrasoRequest {
    fecha: string
    horaEntrada?: string
    horaSalida?: string
    actividad?: string
    evento?: string
    estado: EstadoAsistencia
    observaciones?: string
    responsableId?: number
    responsable?: string
    horaProgramada?: string
}

// Asistencia completa con nuevos campos (del backend)
export interface AsistenciaCompleta {
    id: number
    fecha: string
    horaEntrada?: string
    horaSalida?: string
    actividad?: string
    evento?: string
    estado: string
    observaciones?: string
    latitude?: string
    longitude?: string
    responsableId?: number
    responsable?: string
    createdAt?: string
    // Nuevos campos
    tipoDia?: TipoDia
    tieneRetraso?: boolean
    minutosRetraso?: number
    fechaLimiteRegistro?: string
    horaProgramada?: string
    diaSemana?: number
}

File: backend/src/main/java/com/sidaf/backend/model/Asistencia.java
package com.sidaf.backend.model;

import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "asistencia")
@Data
@NoArgsConstructor
public class Asistencia {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private LocalDate fecha;

    @Column(name = "hora_entrada")
    private LocalDateTime horaEntrada;

    @Column(name = "hora_salida")
    private LocalDateTime horaSalida;

    private String actividad;
    private String evento;
    private String estado;

    @Column(columnDefinition = "text")
    private String observaciones;

    private String latitude;
    private String longitude;

    @Column(name = "responsable_id")
    private Long responsableId;

    private String responsable;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    // ========== NUEVOS CAMPOS PARA MEJORA DE ASISTENCIA ==========

    /**
     * Tipo de día: OBLIGATORIO, OPCIONAL, DESCANSO
     * Días obligatorios: Lunes(1), Martes(2), Jueves(4), Viernes(5), Sábado(6)
     */
    @Column(name = "tipo_dia", length = 20)
    private String tipoDia;

    /**
     * Indica si el registro tiene retraso
     */
    @Column(name = "tiene_retraso")
    private Boolean tieneRetraso = false;

    /**
     * Minutos de retraso (0 si no hay retraso)
     */
    @Column(name = "minutos_retraso")
    private Integer minutosRetraso = 0;

    /**
     * Fecha límite para registrar la asistencia obligatoria
     * Permite registrar con retraso hasta esta fecha
     */
    @Column(name = "fecha_limite_registro")
    private LocalDate fechaLimiteRegistro;

    /**
     * Hora programada de la actividad para calcular retrasos
     */
    @Column(name = "hora_programada")
    private LocalTime horaProgramada;

    /**
     * Día de la semana (1=Lunes, 2=Martes, 3=Miercoles, 4=Jueves, 5=Viernes, 6=Sábado, 7=Domingo)
     * Facilita consultas y filtros por día
     */
    @Column(name = "dia_semana")
    private Integer diaSemana;
}
